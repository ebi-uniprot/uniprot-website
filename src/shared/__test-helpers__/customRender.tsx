/**
 * This tries to follow what is detailed here:
 * https://testing-library.com/docs/react-testing-library/setup#custom-render
 */
import { render, RenderOptions } from '@testing-library/react';
import { Component, Dispatch, ReactElement, ReactNode } from 'react';
import { HelmetProvider } from 'react-helmet-async';
import {
  createMemoryHistory,
  LocationState,
  MemoryHistory,
  Route,
  Router,
} from 'react-router';
import { JsonValue, SetRequired } from 'type-fest';

import { MessagesState } from '../../messages/state/messagesInitialState';
import { MessagesAction } from '../../messages/state/messagesReducers';
import databaseInfo from '../../uniprotkb/utils/__tests__/__mocks__/databaseInfo';
import { getDatabaseInfoMaps } from '../../uniprotkb/utils/database';
import {
  MessagesDispatchContext,
  MessagesStateContext,
} from '../contexts/Messages';
import {
  DatabaseInfoMapsContext,
  UniProtDataVersionContext,
} from '../contexts/UniProtData';
import { JobAction } from '../workers/jobs/state/actionHandler';
import { JobsState } from '../workers/jobs/state/jobsInitialState';

type ExtraRenderOptions = {
  // For react-router
  history?: MemoryHistory<LocationState>;
  path?: string;
  // For custom user preferences (used by useLocalStorage)
  initialLocalStorage?: Record<string, JsonValue>;
  // For app context
  messagesState?: MessagesState;
  messagesDispatch?: jest.Mock<Dispatch<MessagesAction>>;
  JobsState?: JobsState;
  toolsDispatch?: jest.Mock<Dispatch<JobAction>>;
};

type WrapperProps = { children: ReactNode } & RenderOptions &
  SetRequired<
    ExtraRenderOptions,
    | 'history'
    | 'initialLocalStorage'
    | 'messagesState'
    | 'messagesDispatch'
    | 'JobsState'
    | 'toolsDispatch'
  >;

const dbInfoMaps = getDatabaseInfoMaps(databaseInfo);
const uniProtDataVersion = {
  releaseNumber: '2023_01',
  releaseDate: new Date('01-March-2023'),
};

class Wrapper extends Component<WrapperProps> {
  constructor(props: WrapperProps) {
    super(props);
    for (const [key, value] of Object.entries(props.initialLocalStorage)) {
      window.localStorage.setItem(key, JSON.stringify(value));
    }
  }

  // eslint-disable-next-line class-methods-use-this
  componentWillUnmount() {
    window.localStorage.clear();
  }

  render() {
    const { children, path, history, messagesDispatch, messagesState } =
      this.props;
    return (
      <HelmetProvider>
        <Router history={history}>
          <MessagesDispatchContext.Provider value={messagesDispatch}>
            <MessagesStateContext.Provider value={messagesState}>
              <DatabaseInfoMapsContext.Provider value={dbInfoMaps}>
                <UniProtDataVersionContext.Provider value={uniProtDataVersion}>
                  {path ? (
                    <Route path={path} render={() => children} />
                  ) : (
                    children
                  )}
                </UniProtDataVersionContext.Provider>
              </DatabaseInfoMapsContext.Provider>
            </MessagesStateContext.Provider>
          </MessagesDispatchContext.Provider>
        </Router>
      </HelmetProvider>
    );
  }
}

const customRender = (
  ui: ReactElement,
  {
    route = '',
    path,
    history = createMemoryHistory({ initialEntries: [route] }),
    initialLocalStorage = {},
    messagesState = {},
    messagesDispatch = jest.fn(),
    JobsState = {},
    toolsDispatch = jest.fn(),
    ...options
  }: RenderOptions &
    ExtraRenderOptions & {
      route?: string;
    } = {}
) => ({
  ...render(ui, {
    wrapper: (props) => (
      <Wrapper
        path={path}
        history={history}
        initialLocalStorage={initialLocalStorage}
        messagesState={messagesState}
        messagesDispatch={messagesDispatch}
        JobsState={JobsState}
        toolsDispatch={toolsDispatch}
        {...props}
      />
    ),
    ...options,
  }),
  history,
  messagesDispatch,
  toolsDispatch,
});

export default customRender;
