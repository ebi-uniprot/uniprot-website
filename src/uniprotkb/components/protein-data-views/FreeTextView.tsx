import { Fragment, FC, ReactNode } from 'react';
import { Link, useRouteMatch } from 'react-router-dom';

import UniProtKBEvidenceTag from './UniProtKBEvidenceTag';

import {
  getEntryPath,
  getEntryPathFor,
  LocationToPath,
  Location,
  allSearchResultLocations,
} from '../../../app/config/urls';
import {
  reAC,
  needTextProcessingRE,
  rePubMedID,
  rePubMed,
  reUniProtKBAccession,
  reFamily,
  familyExtractor,
  reSubscript,
  reSuperscript,
} from '../../utils';

import { Namespace } from '../../../shared/types/namespaces';
import { FreeTextComment, TextWithEvidence } from '../../types/commentTypes';

import helper from '../../../shared/styles/helper.module.scss';

const needsNewLineRE = /^\)\.\s+/;

const getEntryPathForCitation = getEntryPathFor(Namespace.citations);

export const RichText = ({
  children,
  addPeriod,
}: {
  children?: string;
  addPeriod?: boolean;
}) => (
  <>
    {children?.split(needTextProcessingRE).map((part, index, { length }) => {
      // Capturing group will allow split to conserve that bit in the split parts
      // NOTE: rePubMed and reAC should be using a lookbehind eg `/(?<=pubmed:)(\d{7,8})/i` but
      // it is not supported in Safari yet. It's OK, we just get more chunks when splitting
      const pubMedID = part.match(rePubMedID)?.[0];
      if (rePubMed.test(part) && pubMedID) {
        // PubMed ID, insert a link
        // eg A0A075B6S6
        return (
          // eslint-disable-next-line react/no-array-index-key
          <Fragment key={index}>
            PubMed:
            <Link to={getEntryPathForCitation(pubMedID)}>{pubMedID}</Link>
          </Fragment>
        );
      }
      const accession = part.match(reUniProtKBAccession)?.[0];
      if (reAC.test(part) && accession) {
        // Replace any occurrences of "AC <accession>" with "AC "<link to accession>
        // eg A0A075B6S6
        return (
          // eslint-disable-next-line react/no-array-index-key
          <Fragment key={index}>
            {/* Somehow the part kept "AC ", so put it back */}
            {part.startsWith('AC ') ? `AC ` : ''}
            <Link to={getEntryPath(Namespace.uniprotkb, accession)}>
              {accession}
            </Link>
          </Fragment>
        );
      }
      if (reFamily.test(part)) {
        return part.split(familyExtractor).map((familyPart) => {
          if (familyPart.endsWith('family')) {
            return (
              // eslint-disable-next-line react/no-array-index-key
              <Link
                key={familyPart}
                to={{
                  pathname: LocationToPath[Location.UniProtKBResults],
                  search: `query=(family:"${familyPart}")`,
                }}
              >
                {familyPart}
              </Link>
            );
          }
          return familyPart;
        });
      }
      if (reSubscript.test(part)) {
        return (
          // eslint-disable-next-line react/no-array-index-key
          <sub key={index}>{part.substring(1, part.length - 1)}</sub>
        );
      }
      if (reSuperscript.test(part)) {
        return (
          // eslint-disable-next-line react/no-array-index-key
          <sup key={index}>{part.substring(1, part.length - 1)}</sup>
        );
      }
      // On blocks starting with "). " replace the whitespaces with <br />
      if (needsNewLineRE.test(part)) {
        // add new line before adding the rest of the plain text
        return (
          // eslint-disable-next-line react/no-array-index-key
          <Fragment key={index}>
            ).
            <br />
            {part.replace(needsNewLineRE, '')}
          </Fragment>
        );
      }
      // If the last section doesn't end with a period, add it
      if (addPeriod && index + 1 === length && !part.endsWith('.')) {
        return `${part}.`;
      }
      // use plain text as such
      return part;
    })}
  </>
);

export const TextView = ({ comments }: { comments: TextWithEvidence[] }) => (
  <div className="text-block">
    {comments.map((comment, index) => (
      // eslint-disable-next-line react/no-array-index-key
      <Fragment key={index}>
        <RichText addPeriod>{comment.value}</RichText>
        <UniProtKBEvidenceTag evidences={comment.evidences} />
      </Fragment>
    ))}
  </div>
);

type FreeTextProps = {
  comments?: FreeTextComment[];
  title?: ReactNode;
  articleId?: string;
  showMolecule?: boolean;
};

const FreeTextView: FC<FreeTextProps> = ({
  comments,
  title,
  articleId,
  showMolecule = true,
}) => {
  const searchPageMath = useRouteMatch(allSearchResultLocations);

  if (!comments?.length) {
    return null;
  }

  const freeTextData = comments.map(
    (item, index) =>
      item.texts && (
        // eslint-disable-next-line react/no-array-index-key
        <Fragment key={index}>
          {showMolecule && item.molecule && (
            <h4 className="tiny">
              {searchPageMath?.isExact ? (
                `${item.molecule}`
              ) : (
                <a href={`#${item.molecule.replaceAll(' ', '_')}`}>
                  {item.molecule}
                </a>
              )}
            </h4>
          )}
          <TextView comments={item.texts} />
        </Fragment>
      )
  );

  return (
    <>
      {title && (
        <h3 className={helper.capitalize} data-article-id={articleId}>
          {title}
        </h3>
      )}
      {freeTextData}
    </>
  );
};

export default FreeTextView;
