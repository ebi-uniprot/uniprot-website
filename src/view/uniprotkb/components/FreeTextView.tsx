import React, { Fragment } from 'react';
import { v1 } from 'uuid';
import UniProtEvidenceTag from '../../../components/UniProtEvidenceTag';
import { FreeText } from '../../../model/types/CommentTypes';

type FreeTextProps = {
  comments?: FreeText[];
  includeTitle?: boolean;
};

const FreeTextView: React.FC<FreeTextProps> = ({
  comments,
  includeTitle = false,
}) => {
  if (!comments || comments.length <= 0) {
    return null;
  }
  const freeTextData = comments.map(item => (
    <span className="text-block" key={v1()}>
      {item.texts &&
        item.texts.map(itemText => {
          return (
            <Fragment key={v1()}>
              {itemText.value}
              {itemText.evidences && (
                <UniProtEvidenceTag evidences={itemText.evidences} />
              )}
            </Fragment>
          );
        })}
    </span>
  ));

  return (
    <Fragment>
      {includeTitle && (
        <h4 style={{ textTransform: 'capitalize' }}>
          {comments[0].commentType.toLowerCase()}
        </h4>
      )}
      {freeTextData}
    </Fragment>
  );
};

export default FreeTextView;
