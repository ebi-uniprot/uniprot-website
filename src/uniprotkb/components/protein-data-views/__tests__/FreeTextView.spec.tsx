import { screen } from '@testing-library/react';
import customRender from '../../../../shared/__test-helpers__/customRender';

import FreeTextView from '../FreeTextView';

import freeTextUIData from './__mocks__/freeTextUIData';

describe('FreeText component', () => {
  it('should render free text CC', () => {
    const { asFragment } = customRender(
      <FreeTextView comments={[freeTextUIData[0]]} />
    );
    expect(asFragment()).toMatchSnapshot();

    // PubMed links
    expect(screen.queryAllByRole('link')).toHaveLength(0);

    // Publication button / evidence tag
    expect(
      screen.getByRole('button', { name: '1 publication' })
    ).toBeInTheDocument();
  });

  it('should render free text CC with inlined pubmeds', () => {
    const { asFragment } = customRender(
      <FreeTextView comments={[freeTextUIData[1]]} />
    );
    expect(asFragment()).toMatchSnapshot();

    // PubMed links
    expect(screen.getAllByRole('link')).toHaveLength(2);

    // Publication button / evidence tag
    expect(
      screen.getByRole('button', { name: '1 publication' })
    ).toBeInTheDocument();
  });

  it('should render free text CC, without evidences nor inlined pubmeds', () => {
    const { asFragment } = customRender(
      <FreeTextView comments={[freeTextUIData[0]]} noEvidence />
    );
    expect(asFragment()).toMatchSnapshot();

    // PubMed links
    expect(screen.queryAllByRole('link')).toHaveLength(0);

    // Publication button / evidence tag
    expect(screen.queryByRole('button')).not.toBeInTheDocument();
  });
});
