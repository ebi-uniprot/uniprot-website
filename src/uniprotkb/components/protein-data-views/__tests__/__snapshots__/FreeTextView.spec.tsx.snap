// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FreeText component should render free text CC 1`] = `
<DocumentFragment>
  <div
    class="text-block"
  >
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
    <button
      aria-controls="0"
      aria-expanded="false"
      class="svg-colour-reviewed evidence-tag"
      data-testid="evidence-tag-trigger"
      type="button"
    >
      <test-file-stub
        height="12"
        width="12"
      />
      <span
        class="evidence-tag__label"
      >
        1 publication
      </span>
    </button>
    <div
      class="evidence-tag-content "
      data-testid="evidence-tag-content"
      id="0"
    />
  </div>
</DocumentFragment>
`;

exports[`FreeText component should render free text CC with inlined pubmeds 1`] = `
<DocumentFragment>
  <div
    class="text-block"
  >
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua (PubMed:
    <a
      href="/citations/1234567"
    >
      1234567
    </a>
    , PubMed:
    <a
      href="/citations/12345678"
    >
      12345678
    </a>
    ).
    <br />
    Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
    <button
      aria-controls="0"
      aria-expanded="false"
      class="svg-colour-reviewed evidence-tag"
      data-testid="evidence-tag-trigger"
      type="button"
    >
      <test-file-stub
        height="12"
        width="12"
      />
      <span
        class="evidence-tag__label"
      >
        1 publication
      </span>
    </button>
    <div
      class="evidence-tag-content "
      data-testid="evidence-tag-content"
      id="0"
    />
  </div>
</DocumentFragment>
`;

exports[`FreeText component should render free text CC, without evidences nor inlined pubmeds 1`] = `
<DocumentFragment>
  <div
    class="text-block"
  >
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </div>
</DocumentFragment>
`;
